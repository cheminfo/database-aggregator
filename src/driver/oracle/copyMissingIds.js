'use strict';

const chunk = require('lodash.chunk');

const debug = require('../../util/debug')('driver:oracle');
const model = require('../../mongo/model');
const sqlUtil = require('../../util/sql');

const common = require('./common');

const MAX_ELEMENTS_ID_CLAUSE = 999;

// In a similar fashion to how remove works, finds all the ids
// that are present in the source but missing in the target
// and copies those to the target

module.exports = function (options) {
  common.checkOptions(['connectString', 'query'], options);
  return doSync(options);
};

async function doSync(options) {
  const collection = options.collection;
  const Model = model.getSource(collection);

  // connect to Oracle
  const oracleConn = await common.connect(options);

  const query = options.query;
  const idQuery = `SELECT id FROM (\n${query}\n) inner_table`;

  const latest = await Model.findOne()
    .sort('-date')
    .exec();

  // Get all ids from original source
  const sourceIds = await common.getIDs(oracleConn, idQuery);

  // Get all ids from copied source
  const targetIds = new Set(
    (await Model.find({ data: { $ne: null } }, { _id: 1 })
      .lean()
      .exec()).map((t) => t._id)
  );
  const idsToCopy = new Set();

  for (let id of sourceIds) {
    // add to list if exists in original source but not in copied source
    if (!targetIds.has(id)) {
      idsToCopy.add(id);
    }
  }

  debug(`adding ${idsToCopy.size} missing entries`);
  const chunks = chunk([...idsToCopy], MAX_ELEMENTS_ID_CLAUSE);
  for (let chunk of chunks) {
    let copyQuery = `
            SELECT * FROM (
                ${query}
            ) inner_table
            WHERE moddate < ${sqlUtil.formatTimestamp(latest.date)}
            AND ID IN ('${chunk.join("','")}')
        `;
    await common.copyEntries(oracleConn, copyQuery, collection);
  }

  await oracleConn.release();
}
