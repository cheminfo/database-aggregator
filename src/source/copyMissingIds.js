'use strict';

const chunk = require('lodash.chunk');

const model = require('../mongo/model');
const { connect } = require('../mongo/connection');
const debug = require('../util/debug')('source:copyMissingIds');

const getDriver = require('./getDriver');
const copyEntries = require('./copyEntries');

const MAX_ELEMENTS_ID_CLAUSE = 999;

// In a similar fashion to how remove works, finds all the ids
// that are present in the source but missing in the target
// and copies those to the target
async function copyMissingIds(config) {
  const driver = getDriver(config.driver);

  let sourceIds = await driver.getIds(config);
  if (!(sourceIds instanceof Set)) {
    sourceIds = new Set(sourceIds);
  }

  const collection = config.collection;
  const Model = model.getSource(collection);

  await connect();

  // Get all ids from copied source
  const targetIds = new Set(
    (await Model.find({ data: { $ne: null } }, { id: 1 })
      .lean()
      .exec()).map((t) => t.id)
  );
  const idsToCopy = new Set();

  for (let id of sourceIds) {
    // add to list if exists in original source but not in copied source
    if (!targetIds.has(id)) {
      idsToCopy.add(id);
    }
  }

  const latest = await Model.findOne()
    .sort('-date')
    .exec();

  debug(`adding ${idsToCopy.size} missing entries`);
  const chunks = chunk([...idsToCopy], MAX_ELEMENTS_ID_CLAUSE);

  for (let chunk of chunks) {
    await driver.getData(config, (data) => copyEntries(data, config), {
      latestDate: (latest && latest.date) || new Date('1900-01-01'),
      ids: chunk
    });
  }
}

module.exports = copyMissingIds;
