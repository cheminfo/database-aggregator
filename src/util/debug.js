'use strict';

const Debug = require('debug');

var error = Debug('aggregate:error');
var warn = Debug('aggregate:warn');
var debug = Debug('aggregate:debug');
var trace = Debug('aggregate:trace');

warn.log = console.warn.bind(console);
debug.log = console.log.bind(console);
trace.log = console.log.bind(console);

module.exports = function (prefix) {
  const func = (message) => debug(`(${prefix}) ${message}`);
  func.error = (message) => error(`(${prefix}) ${message}`);
  func.warn = (message) => warn(`(${prefix}) ${message}`);
  func.debug = func;
  func.trace = (message) => trace(`(${prefix}) ${message}`);
  return func;
};
