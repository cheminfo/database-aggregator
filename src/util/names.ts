const AGGREGATION = 'aggregation';
const SOURCE = 'source';
const META = 'meta';

const REMOVE_ID = 'source_remove_';
const COPY_ID = 'source_copy_';
const COPY_MISSING_ID = 'source_copy_missing_ids_';

export function getModelName(prefix: string, name: string) {
  return `${prefix}_${name}`;
}

export function getSourceName(name: string) {
  return getModelName(SOURCE, name);
}

export function getMetaModelName(name: string) {
  return getModelName(META, name);
}

export function getSourceModelName(name: string) {
  return getModelName(SOURCE, name);
}

export function getAggregationModelName(name: string) {
  return getModelName(AGGREGATION, name);
}

export function getCopyTaskId(collection: string) {
  return COPY_ID + collection;
}

export function getRemoveTaskId(collection: string) {
  return REMOVE_ID + collection;
}

export function getCopyMissingIdTaskId(collection: string) {
  return COPY_MISSING_ID + collection;
}
