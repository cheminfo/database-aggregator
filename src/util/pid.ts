/* eslint-disable no-process-exit */

const path = require("path");
const fs = require("fs");

const isRunning = require("is-running");

const pidFile = path.join(__dirname, "../../source.pid");

export function start() {
  try {
    const pid = fs.readFileSync(pidFile, "utf8");
    if (isRunning(pid)) {
      console.log(`process is running (pid: ${pid})`);
      process.exit(0);
    }
    console.log("pid file is here but process died");
    fs.unlinkSync(pidFile);
  } catch (e) {
    if (e.code === "ENOENT") {
      console.log("pid file does not exist yet");
    } else {
      throw e;
    }
  }
  const currentPid = process.pid;
  fs.writeFileSync(pidFile, String(currentPid));
}

export function stop(code?: number) {
  code = code || 0;
  fs.unlinkSync(pidFile);
  process.exit(code);
}
