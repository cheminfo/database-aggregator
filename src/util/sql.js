'use strict';

// http://stackoverflow.com/a/5133807
function twoDigits(d) {
  if (d >= 0 && d < 10) return `0${d.toString()}`;
  if (d > -10 && d < 0) return `-0${(-1 * d).toString()}`;
  return d.toString();
}

// exports.formatDate = function (date) {
//    return `TO_DATE('${date.getUTCFullYear()}-${twoDigits(1 + date.getUTCMonth())}-${twoDigits(date.getUTCDate())}', 'YYYY-MM-DD')`;
// };

// exports.formatDateTime = function (date) {
//    return `TO_DATE('${date.getUTCFullYear()}-${twoDigits(1 + date.getUTCMonth())}-${twoDigits(date.getUTCDate())} ${twoDigits(date.getUTCHours())}:${twoDigits(date.getUTCMinutes())}:${twoDigits(date.getUTCSeconds())}', 'YYYY-MM-DD HH24:MI:SS')`;
// };

exports.formatTimestamp = function (date) {
  return `(TIMESTAMP '${date.getUTCFullYear()}-${twoDigits(1 + date.getUTCMonth())}-${twoDigits(date.getUTCDate())} ${twoDigits(date.getUTCHours())}:${twoDigits(date.getUTCMinutes())}:${twoDigits(date.getUTCSeconds())} UTC' AT TIME ZONE SESSIONTIMEZONE)`;
};
