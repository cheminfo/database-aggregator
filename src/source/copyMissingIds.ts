const chunkLib = require('lodash.chunk');

import { connect } from '../mongo/connection';
import { getSource } from '../mongo/model';

import { ISourceConfigElement } from '../types';
import { debugUtil } from '../util/debug';
import { copyEntries } from './copyEntries';
import { getDriver } from './getDriver';
import { ISourceDocument } from '../internalTypes';

const debug = debugUtil('source:copyMissingIds');

const MAX_ELEMENTS_ID_CLAUSE = 999;

// In a similar fashion to how remove works, finds all the ids
// that are present in the source but missing in the target
// and copies those to the target
export async function copyMissingIds(config: ISourceConfigElement) {
  const driver = getDriver(config.driver);

  let sourceIds = await driver.getIds(config.driverConfig);
  if (!(sourceIds instanceof Set)) {
    sourceIds = new Set(sourceIds);
  }

  const collection = config.collection;
  const Model = getSource(collection);

  await connect();

  // Get all ids from copied source
  const targetIds: Set<string> = new Set(
    (await Model.find({ data: { $ne: null } }, { id: 1 }).lean()).map(
      (t: ISourceDocument) => t.id
    )
  );
  const idsToCopy = new Set();

  for (const id of sourceIds) {
    // add to list if exists in original source but not in copied source
    if (!targetIds.has(id)) {
      idsToCopy.add(id);
    }
  }

  const latest = await Model.findOne()
    .sort('-date')
    .exec();

  debug.debug(`adding ${idsToCopy.size} missing entries`);
  const chunks = chunkLib([...idsToCopy], MAX_ELEMENTS_ID_CLAUSE);

  for (const chunk of chunks) {
    await driver.getData(
      config.driverConfig,
      (data) => copyEntries(data, config),
      {
        latestDate: (latest && latest.date) || new Date('1900-01-01'),
        ids: chunk
      }
    );
  }
}
