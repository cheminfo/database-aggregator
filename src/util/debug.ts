import * as Debug from "debug";

let error = Debug("aggregator:error");
let warn = Debug("aggregator:warn");
let debug = Debug("aggregator:debug");
let trace = Debug("aggregator:trace");

warn.log = console.warn.bind(console);
debug.log = console.log.bind(console);
trace.log = console.log.bind(console);

export function debugUtil(prefix: string) {
  return {
    error: (message: string) => error(`(${prefix}) ${message}`),
    warn: (message: string) => warn(`(${prefix}) ${message}`),
    debug: (message: string) => debug(`(${prefix}) ${message}`),
    trace: (message: string) => trace(`(${prefix}) ${message}`),
  };
}
