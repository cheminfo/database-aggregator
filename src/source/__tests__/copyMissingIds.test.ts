import {
  connect,
  disconnect,
  dropDatabase,
  insertData
} from '../../../test/mongoSetup';
import { getCollection, clean } from '../../../test/util';
import { copyMissingIds } from '../copyMissingIds';

beforeAll(connect);
afterAll(disconnect);

const collection = getCollection('source_test');
const config = {
  driver: {
    getIds(config) {
      const data = ['test1', 'test2', 'test3'];
      if (config.type === 'set') return new Set(data);
      return data;
    },
    getData(config, callback, options) {
      if (!options.ids) {
        throw new Error('unexpected');
      }
      if (options.ids.length > 1) {
        throw new Error('unexpected');
      }
      return callback([
        {
          id: 'test3',
          commonID: 'test3',
          modificationDate: new Date(0),
          data: { test: 3 }
        }
      ]);
    }
  },
  collection: 'test'
};

const testData = {
  // eslint-disable-next-line camelcase
  meta_source_sequence: [
    {
      _id: 'test',
      seq: 2
    }
  ],
  // eslint-disable-next-line camelcase
  source_test: [
    {
      id: 'test1',
      commonID: 'test1',
      date: new Date(0),
      sequentialID: 1,
      data: {}
    },
    {
      id: 'test2',
      commonID: 'test2',
      date: new Date(0),
      sequentialID: 2,
      data: {}
    }
  ]
};

describe('source copyMissingIds', () => {
  beforeEach(async () => {
    await dropDatabase();
    await insertData(testData);
  });
  it('should copy missing entries', async () => {
    await copyMissingIds(config);
    const data = await collection.find().toArray();
    expect(data[2].data).toEqual({ test: 3 });
    expect(clean(data)).toMatchSnapshot();
  });

  it('should copy missing entries when driver returns a set', async () => {
    await copyMissingIds(Object.assign(config, { type: 'set' }));
    const data = await collection.find().toArray();
    expect(data[2].data).toEqual({ test: 3 });
    expect(clean(data)).toMatchSnapshot();
  });
});
