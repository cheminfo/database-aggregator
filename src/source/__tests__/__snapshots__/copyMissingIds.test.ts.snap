// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`source copyMissingIds should copy missing entries 1`] = `
Array [
  Object {
    "commonID": "test1",
    "data": Object {},
    "date": 1970-01-01T00:00:00.000Z,
    "id": "test1",
    "sequentialID": 1,
  },
  Object {
    "commonID": "test2",
    "data": Object {},
    "date": 1970-01-01T00:00:00.000Z,
    "id": "test2",
    "sequentialID": 2,
  },
  Object {
    "__v": 0,
    "commonID": "test3",
    "data": Object {
      "test": 3,
    },
    "date": 1970-01-01T00:00:00.000Z,
    "id": "test3",
    "sequentialID": 3,
  },
]
`;

exports[`source copyMissingIds should copy missing entries when driver returns a set 1`] = `
Array [
  Object {
    "commonID": "test1",
    "data": Object {},
    "date": 1970-01-01T00:00:00.000Z,
    "id": "test1",
    "sequentialID": 1,
  },
  Object {
    "commonID": "test2",
    "data": Object {},
    "date": 1970-01-01T00:00:00.000Z,
    "id": "test2",
    "sequentialID": 2,
  },
  Object {
    "__v": 0,
    "commonID": "test3",
    "data": Object {
      "test": 3,
    },
    "date": 1970-01-01T00:00:00.000Z,
    "id": "test3",
    "sequentialID": 3,
  },
]
`;
